### PR TITLE
Openstack

### PR DESCRIPTION
**Pull request by enolfc**
# Summary 

Openstack section update. Brought contents from the wiki and reorganised things a bit

### CODE REVIEW

**Comment by enolfc** [comment]
This is to be rebased

File: source/check-in.rst [hunk 1]
```diff
@@ -1,26 +1,40 @@
...
-Obtaining an access token
-^^^^^^^^^^^^^^^^^^^^^^^^^
+OpenID Connect using Check-in
+-----------------------------
+
+`OpenID Connect <http://openid.net/connect/>`_ is an authentication protocol
+based on OAuth 2.0 and replacing the current VOMS-based authentication on the
+EGI Cloud.
+
+Those providers natively integrated with EGI Check-in are able to authenticate
```
Comment (gwarf) [inline source/check-in.rst]:
```suggestion
The providers natively integrated with EGI Check-in are able to authenticate
```

File: source/check-in.rst [hunk 2]
```diff
@@ -1,26 +1,40 @@
...
+users with OAuth 2.0 tokens without using any certificates in the process by
+relying on the OpenStack Keystone `OS-FEDERATION API <https://developer.openstack.org/api-ref/identity/v3-ext/index.html#os-federation-api>`_.
+
+.. TODO(enolfc): discovery of providers with check-in support?
+
+The process for authentication is as follows:
+
+#. Obtain a valid access token from Check-in. Access tokens are short-lived
+   credentials that can be obtanied by recognised Check-in clients once a user
```
Comment (gwarf) [inline source/check-in.rst]:
```suggestion
   credentials that can be obtained by recognised Check-in clients once a user
```

File: source/check-in.rst [hunk 3]
```diff
@@ -1,26 +1,40 @@
...
+   has authenticated.
```
Comment (gwarf) [inline source/check-in.rst]:
```suggestion
   has been authenticated.
```

File: source/check-in.rst [hunk 4]
```diff
@@ -41,20 +55,221 @@ Check-in credentials to obtain:
...
+Your access token will provide you access to a cloud provider, but you may have
+access to several different projects within that provider (a project can be
+considered as equal to a VO allocation). In order to discover which projects
+are available you can do that using the Keystone API.
+
+The :download:`../get-projects.py` script will just do that for you (requires
+``requests`` library). It expects these variables to be defined in the
+environment:
+
+* ``CHECKIN_CLIENT_ID``: Your Check-in client id (get it from https://aai.egi.eu/fedcloud)
```
Comment (gwarf) [inline source/check-in.rst]: May want to use properly declared links.

File: source/check-in.rst [hunk 5]
```diff
@@ -41,20 +55,221 @@ Check-in credentials to obtain:
...
+
+     cat /etc/grid-security/certificates/*.pem >> $(python -m requests.certs)
+
+
+Access the provider
+^^^^^^^^^^^^^^^^^^^
+
+Once you know which project to use, you can get use your regular openstack cli
+commands for performing actual operations in the provider:
+
```
Comment (gwarf) [inline source/check-in.rst:141]:
```suggestion
```

File: source/check-in.rst [hunk 6]
```diff
@@ -41,20 +55,221 @@ Check-in credentials to obtain:
...
+   .. code-block:: console
+
+     cat /etc/grid-security/certificates/*.pem >> $(python -m requests.certs)
+
+
+Access the provider
+^^^^^^^^^^^^^^^^^^^
+
+Once you know which project to use, you can get use your regular openstack cli
```
Comment (gwarf) [inline source/check-in.rst]:
```suggestion
Once you know which project to use, you can use your regular openstack cli
```

File: source/check-in.rst [hunk 7]
```diff
@@ -41,20 +55,221 @@ Check-in credentials to obtain:
...
+   export OS_ACCESS_TOKEN=$(curl -s -X POST -u "$CHECKIN_CLIENT_ID":"$CHECKIN_CLIENT_SECRET"  \
+        -d "client_id=$CHECKIN_CLIENT_ID&$CHECKIN_CLIENT_SECRET&grant_type=refresh_token&refresh_token=$CHECKIN_REFRESH_TOKEN&scope=openid%20email%20profile" \
+       'https://aai.egi.eu/oidc/token' | jq -r '.access_token')
+   openstack image list
+
+.. note::
+
+   For 3rd party tools that can use token based authentication in OpenStack,
+   use the following command (after setting the environment as shown above):
+
```
Comment (gwarf) [inline source/check-in.rst:165]:
```suggestion
```

File: source/check-in.rst [hunk 8]
```diff
@@ -41,20 +55,221 @@ Check-in credentials to obtain:
...
+^^^^^^^^^^^^^^^^^^
+
+Every VO needs two different pieces of information:
+
+* the ``vomses`` configuration files, where the details of the VO are stored
+  (e.g. name, server, ports). These are stored by default at ``/etc/vomses``
+  and are normally named following this convention: ``<vo name>.<server name>``
+  (e.g. for fedcloud.egi.eu VO, you would have ``fedcloud.egi.eu.voms1.grid.cesnet.cz``
+  and ``fedcloud.egi.eu.voms2.grid.cesnet.cz``.
+
```
Comment (gwarf) [inline source/check-in.rst:191]:
```suggestion
```

File: source/index.rst [hunk 9]
```diff
@@ -5,6 +5,14 @@
 EGI Federated Cloud
 ===================
 
+The EGI Cloud service offers a multi-cloud IaaS federation that brings together
+research clouds as a scalable computing platform for data and/or compute
+driven applications and services in research and science.
```
Comment (gwarf) [inline source/index.rst]:
```suggestion
driven applications and services for research and science.
```

File: source/index.rst [hunk 10]
```diff
@@ -5,6 +5,14 @@
...
+
+This documentation focuses on using the service. Those resource providers
+willing to integrate into the service, plase check `EGI Federated Cloud Integration documentation <https://egi-federated-cloud-integration.readthedocs.io>`_.
+
```
Comment (gwarf) [inline source/index.rst:14]:
```suggestion
```

File: source/introduction.rst [hunk 11]
```diff
@@ -1,7 +1,128 @@
...
+
+* Resource discovery features to easily understand which providers are support
+  your community and what are their capabilities.
+
+* `Global accounting <https://accounting.egi.eu/cloud/>`_ that aggregates and
+  allows visualisation of usage information across the whole federation.
+
+* `Monitoring of Availability and Reliability of the providers <http://argo.egi.eu/status-fedcloud>`_
+  to ensure SLAs are met.
+
```
Comment (gwarf) [inline source/introduction.rst:32]:
```suggestion
```

File: source/introduction.rst [hunk 12]
```diff
@@ -1,7 +1,128 @@
...
+    `The Chipster Platform <https://www.egi.eu/use-cases/research-stories/new-viruses-implicated-in-fatal-snake-disease/>`_.
+
+* **Datasets repository**: the EGI Cloud can be used to store and manage large
+  datasets exploiting the big amount of disk storage available in the Federation.
+
+* **Disposable and testing environments**: environments for training or testing
+  new developments.
+
+  * Example: `Training infrastructure <https://wiki.egi.eu/wiki/Training_infrastructure>`_.
+
```
Comment (gwarf) [inline source/introduction.rst:67]:
```suggestion
```

File: source/introduction.rst [hunk 13]
```diff
@@ -1,7 +1,128 @@
...
+  resources in term of computation and/or memory and/or intensive I/O. Ad-hoc
+  computing environments can be created in the EGI cloud providers to satisfy
+  very hard HW resource requirements.
+
+  * Example: `VERCE platform <https://www.egi.eu/news/new-egi-use-case-a-close-look-at-the-amatrice-earthquake/>`_,
+    `The Genetics of Salmonella Infections <https://www.egi.eu/use-cases/research-stories/the-genetics-of-salmonella-infections/>`_,
+    `The Chipster Platform <https://www.egi.eu/use-cases/research-stories/new-viruses-implicated-in-fatal-snake-disease/>`_.
+
+* **Datasets repository**: the EGI Cloud can be used to store and manage large
+  datasets exploiting the big amount of disk storage available in the Federation.
```
Comment (gwarf) [inline source/introduction.rst]:
```suggestion
  datasets exploiting the large amount of disk storage available in the Federation.
```

File: source/introduction.rst [hunk 14]
```diff
@@ -1,7 +1,128 @@
...
+* **Disposable and testing environments**: environments for training or testing
+  new developments.
+
+  * Example: `Training infrastructure <https://wiki.egi.eu/wiki/Training_infrastructure>`_.
+
+
+Getting access
+--------------
+
+
```
Comment (gwarf) [inline source/introduction.rst]:
```suggestion
```

File: source/introduction.rst [hunk 15]
```diff
@@ -1,7 +1,128 @@
...
+authentication. While we are transitioning to a certificate-less experience for
+using the service, membership to the VO still require users to obtain a personal
+certificate from a recognised Certification Authoriy (unless you have one
+already)
+
+Obtaining a certificate
+^^^^^^^^^^^^^^^^^^^^^^^
+
+The easiest option is to get an 'eScience Personal' certificate online from the
+Terena Certificate Service CA. Check here the countries where this is available,
```
Comment (gwarf) [inline source/introduction.rst]:
```suggestion
Terena Certificate Service CA. Check the countries where this is available,
```

File: source/introduction.rst [hunk 16]
```diff
@@ -1,7 +1,128 @@
...
+and follow the link to the respective CA page at the `TCS participants <https://www.terena.org/activities/tcs/participants.html>`_
+list (See FAQs for details.)
+
+If eScience Personal certificate is not available in your country, then obtain
+a certificate from a regular `https://www.eugridpma.org/members/worldmap/ <IGTF CA>`_
+(this requires personal visit at the CA).
```
Comment (gwarf) [inline source/introduction.rst]:
```suggestion
(this may require a personal visit at the CA).
```

File: source/introduction.rst [hunk 17]
```diff
@@ -1,7 +1,128 @@
...
+
+.. warning::
+
+  * After the 6-month long membership in the fedcloud.egi.eu VO, you will need
+    to move to a production VO, or establish a new VO.
+
+  * The resources are not guaranteed and may be removed without notice by
+    providers. **Back-up frequently to avoid loosing your work!**
+
+
```
Comment (gwarf) [inline source/introduction.rst]:
```suggestion
```

File: source/introduction.rst [hunk 18]
```diff
@@ -1,7 +1,128 @@
...
+Other VOs
+^^^^^^^^^
+
+Pre-existing VOs of EGI can be also used on IaaS cloud providers. Consult with
+your VO manager.  If none of the existing VOs matches your use case, then a
```
Comment (gwarf) [inline source/introduction.rst]:
```suggestion
your VO manager. If none of the existing VOs matches your use case, then a
```

File: source/usage-openstack.rst [hunk 19]
```diff
@@ -1,97 +1,78 @@
...
 
-.. code-block:: console
+* glance, for VM image management
 
-   cat /etc/grid-security/certificates/*.pem >> $(python -m requests.certs)
+* cinder, for block storage
 
-Installing requirements
-^^^^^^^^^^^^^^^^^^^^^^^
+* neutron, for network management, and
```
Comment (gwarf) [inline source/usage-openstack.rst]:
```suggestion
* neutron, for network management
```

File: source/usage-openstack.rst [hunk 20]
```diff
@@ -100,13 +81,13 @@ You can refresh the access token and obtain an OpenStack token in a single
 :download:`../get-token.py` script expecting your credentials to be available in
 the environment:
 
-* CHECKIN_CLIENT_ID: Your Check-in client id (get it from https://aai.egi.eu/fedcloud)
-* CHECKIN_CLIENT_SECRET: Your Check-in client secret (get it from https://aai.egi.eu/fedcloud)
-* CHECKIN_REFRESH_TOKEN: Your Check-in refresh token (get it from https://aai.egi.eu/fedcloud)
-* OS_AUTH_URL: Keystone URL (depends on the provider, you can get it in https://goc.egi.eu)
-* OS_PROJECT_ID: OpenStack project to use (See script above for obtaining it)
+* ``CHECKIN_CLIENT_ID``: Your Check-in client id (get it from https://aai.egi.eu/fedcloud)
```
Comment (gwarf) [inline source/usage-openstack.rst]: Here also using proper links may be better.

File: source/usage-openstack.rst [hunk 21]
```diff
@@ -201,7 +188,167 @@ cloud-config data as user data
 * `Official cloud-config examples <https://cloudinit.readthedocs.io/en/latest/topics/examples.html#yaml-examples>`_
 * `Cloud-init example <https://www.zetta.io/en/help/articles-tutorials/customizing-instance-deployment-cloud-init/>`_
 
-Other resources
----------------
+Terraform
+---------
+
+`Terraform <https://terraform.io>`_  supports EGI Cloud OpenStack providers by
```
Comment (gwarf) [inline source/usage-openstack.rst]:
```suggestion
`Terraform <https://terraform.io>`_ supports EGI Cloud OpenStack providers by
```

File: source/usage-openstack.rst [hunk 22]
```diff
@@ -201,7 +188,167 @@ cloud-config data as user data
...
+   ------------------------------------------------------------------------
+
+   Note: You didn't specify an "-out" parameter to save this plan, so Terraform
+   can't guarantee that exactly these actions will be performed if
+   "terraform apply" is subsequently run.
+
+Note that as in the example above you can get images using information from
+AppDB if needed.
+
+
```
Comment (gwarf) [inline source/usage-openstack.rst]:
```suggestion
```

**Review by gwarf — CHANGES_REQUESTED** [review]
Nice, great to see all this doc, well done!
Just a few cosmetics sorry for being so psycho-rigid with spaces/lines. (I wonder if one empty line after a code block wouldn't be enough too).

File: source/check-in.rst (hunk 4 above)
Comment (enolfc) [reply source/check-in.rst]: Fixed

File: source/usage-openstack.rst (hunk 20 above)
Comment (enolfc) [reply source/usage-openstack.rst]: Also fixed

**Review by gwarf — APPROVED** [review]

**Pull request merged**